c2rust-analyze/src/main.rs

```diff
@@ -35,29 +41,113 @@ mod pointer_id;
 mod type_desc;
 mod util;
 
+/// A wrapper around `T` that dynamically tracks whether it's initialized or not.  `RefCell`
+/// dynamically tracks borrowing and panics if the rules are violated at run time; `MaybeUnset`
+/// dynamically tracks initialization and similarly panics if the value is accessed while unset.
+#[derive(Clone, Copy, Debug)]
+struct MaybeUnset<T>(Option<T>);
+
+impl<T> Default for MaybeUnset<T> {
+    fn default() -> MaybeUnset<T> {
+        MaybeUnset(None)
+    }
+}
+
+impl<T> MaybeUnset<T> {
+    pub fn set(&mut self, x: T) {
+        if self.0.is_some() {
+            panic!("value is already set");
+        }
+        self.0 = Some(x);
+    }
+
+    pub fn clear(&mut self) {
+        if self.0.is_none() {
+            panic!("value is already cleared");
+        }
+        self.0 = None;
+    }
+
+    pub fn get(&self) -> &T {
+        self.0.as_ref().expect("value is not set")
+    }
+
+    pub fn get_mut(&mut self) -> &mut T {
+        self.0.as_mut().expect("value is not set")
+    }
+
+    pub fn take(&mut self) -> T {
+        self.0.take().expect("value is not set")
+    }
+}
+
+impl<T> Deref for MaybeUnset<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        self.get()
+    }
+}
+
+impl<T> DerefMut for MaybeUnset<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        self.get_mut()
+    }
+}
+
 fn run(tcx: TyCtxt) {
     let mut gacx = GlobalAnalysisCtxt::new(tcx);
-    let mut func_info = Vec::new();
+    let mut func_info = HashMap::new();
+
+    #[derive(Default)]
+    struct FuncInfo<'tcx> {
+        acx_data: MaybeUnset<AnalysisCtxtData<'tcx>>,
+        dataflow: MaybeUnset<DataflowConstraints>,
+        l_equiv: MaybeUnset<LocalEquivSet>,
```

Comment:
Previously `l` meant `Labeled`.  Here I assume it means `Local`.  I wouldn't use `l`/`L` for both of them.

Reply:
OK, I renamed all the `l_`/`g_`s to `local_`/`global_`